### PR TITLE
Add option to authenticate with Azure AD.

### PR DESCRIPTION
**Pull request by askew**
This is quite a big change.

It adds the option to use [Azure Active Directory (Azure AD)][aad] instead of the username/password stored in the database.

There is a new option `auth` that can be set to `db` (the default) or `azuread`.

It uses the [Microsoft Authentication Library (MSAL)][msal] to handle the redirect using the [OpenID Connect (OIDC) authorisation code flow][codeflow].

Entries are still added to the user table using object identifier from Azure AD to match to the `uuid` field in the database.

Admin permission is now governed by an [AppRole][approle] in Azure AD. Admins can then be managed centrally. The admin permission is updated in the database on log in to ensure consistency with rest of the application.

To configure the application for Azure AD authentication the following options need to be set:
  * `auth` = `azuread`
  * `client_id` = the client id/app id of the application registration in Azure AD
  * `tenant_id` = the identifier for the Azure AD tenant (a GUID)
  * `client_secret` = the secret key in Azure AD used to authenticate the application with Azure AD.
  * `redirect_url` = the fully qualified URL that Azure AD will redirect to after the user signs in.

Note, the `redirect_url` default to http://localhost:8888/oidc. The /oidc path is handled by the new `CodeFlowHandler` in `PublicHandlers.py`.

Registration is disables in `azuread` mode, and a new 'Join Team' page allows a new user to join an existing team.

[aad]: https://azure.microsoft.com/services/active-directory/ "Azure Active Directory"
[msal]: https://github.com/AzureAD/microsoft-authentication-library-for-python "Microsoft Authentication Library (MSAL) for Python"
[approle]: https://go.microsoft.com/fwlink/?linkid=2138582 "Assign users and groups to roles"
[codeflow]: https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-auth-code-flow#request-an-authorization-code "Microsoft identity platform and OAuth 2.0 authorization code flow"

### CODE REVIEW

**Comment by anthturner** [comment]
+1

**Comment by jeffg2k** [comment]
Nice - this should also help #240

File: handlers/PublicHandlers.py [hunk 1]
```diff
@@ -61,16 +65,149 @@ def get(self, *args, **kwargs):
...
+            elif team_code is None or len(team_code) == 0:
+                needsateam = True
+        return needsateam
+
+    def add_user(self, claims, team_code):
+        user = User()
+        user.uuid = claims["oid"]
+        user.handle = claims["preferred_username"].split("@")[0]
+        # Generate a long random password that the user will never know or use.
+        user.password = ''.join(random.choices(string.ascii_letters + string.digits + string.punctuation, k=30))
```
Comment (moloch--) [inline handlers/PublicHandlers.py:166]: Very minor nitpick -- I would encourage the use of [`secrets`](https://docs.python.org/3/library/secrets.html?highlight=secrets#module-secrets) or [`SystemRandom`](https://docs.python.org/3/library/random.html#random.SystemRandom) here for password generation as `random` is not cryptographically secure.

Comment (askew) [reply handlers/PublicHandlers.py]: Changed to use `secrets`.

**Comment by jeffg2k** [comment]
@askew I added your description to the Wiki https://github.com/moloch--/RootTheBox/wiki/Cloud-Deployment#azure-ad
Feel free to update or expand on it to better explain implementation to others.  Thanks Again

**Pull request merged**